test/exercises/diffusion/heat_2d_dist_buffers.chpl

```diff
@@ -0,0 +1,149 @@
+/*
+  A distributed 2D finite-difference heat/diffusion equation solver
+
+  Computation is executed over a 2D distributed array.
+  The array distribution is managed by the `Block` distribution.
+  Tasks are spawned manually with a `coforall` loop and synchronization
+  is done manually using a `barrier`. Halo regions are shared across
+  locales manually via halo/buffer arrays.
+*/
+
+import BlockDist.Block,
+       Collectives.barrier,
+       Time.stopwatch;
+
+// create a stopwatch to time kernel execution
+var t = new stopwatch();
+config const writeTime = false;
+
+// compile with `-sRunCommDiag=true` to see comm diagnostics
+use CommDiagnostics;
+config param RunCommDiag = false;
+
+// declare configurable constants with default values
+config const nx = 256,      // number of grid points in x
+             ny = 256,      // number of grid points in y
+             nt = 50,       // number of time steps
+             alpha = 0.25;  // diffusion constant
+
+// define distributed domains and block-distributed array
+const indices = {0..<nx, 0..<ny},
+      indicesInner = indices.expand(-1),
+      INDICES = Block.createDomain(indices);
+var u: [INDICES] real;
+
+// apply initial conditions
+u = 1.0;
+u[nx/4..nx/2, ny/4..ny/2] = 2.0;
+
+// a type for creating a "skyline" array of halo buffers
+record haloArray {
+  var d: domain(1);
+  var v: [d] real;
+
+  proc init() do this.d = {0..0};
+  proc init(r: range(int)) do
+    this.d = {r};
+}
+
+// set up array of halo buffers over same distribution as 'u.targetLocales'
+var LOCALE_DOM = Block.createDomain(u.targetLocales().domain);
+var haloArrays: [LOCALE_DOM] [0..<4] haloArray;
+
+// buffer edge indices: North, East, South, West
+param N = 0, S = 1, E = 2, W = 3;
+
+// number of tasks per dimension based on Block distribution's decomposition
+const tidXMax = u.targetLocales().dim(0).high,
+      tidYMax = u.targetLocales().dim(1).high;
+
+// barrier for one task per locale
+var b = new barrier(numLocales);
```

Comment:
This also seems indirect—I'd define it in terms of `LOCALE_SPACE.size` since that's what's controlling the number of things you're barriering between.